src/Model.hs

```diff
@@ -97,237 +77,191 @@ handleCore m env = (handleDist . handleRead env) (runModel m)
+envDraw :: forall env sig m a x. (Observable env x a, Has (ObsReader env :+: Draw) sig m)
+  => PrimDist a
   -> ObsVar x
-  -> Model env es a
-deterministic x field = Model $ do
+  -> m a
+envDraw primDist field = do
   let tag = Just $ varToStr field
   maybe_y <- ask @env field
-  call (Dist (DeterministicDist x) maybe_y tag)
+  draw primDist maybe_y tag
+
+deterministic :: forall env sig m a x. (Eq a, Show a, OpenSum.Member a PrimVal, Observable env x a, Has (ObsReader env :+: Draw) sig m)
+  => a
+  -> ObsVar x
+  -> m a
+deterministic x = envDraw @env (DeterministicDist x)
 
-deterministic' :: (Eq a, Show a, OpenSum.Member a PrimVal) =>
-     a -- ^ value to be deterministically generated
-  -> Model env es a
-deterministic' x = Model $ do
-  call (Dist (DeterministicDist x) Nothing Nothing)
+deterministic' :: (Eq a, Show a, OpenSum.Member a PrimVal, Has (ObsReader env :+: Draw) sig m)
+  => a -- ^ value to be deterministically generated
+  -> m a
+deterministic' x =
+  draw (DeterministicDist x) Nothing Nothing
 
-dirichlet :: forall env es x. Observable env x [Double] =>
-     [Double]
+dirichlet :: forall env sig m x. (Observable env x [Double], Has (ObsReader env :+: Draw) sig m)
+  => [Double]
   -> ObsVar x
-  -> Model env es [Double]
-dirichlet xs field = Model $ do
-  let tag = Just $ varToStr field
-  maybe_y <- ask @env field
-  call (Dist (DirichletDist xs) maybe_y tag)
+  -> m [Double]
+dirichlet xs = envDraw @env (DirichletDist xs)
 
-dirichlet' ::
-     [Double] -- ^ concentration parameters
-  -> Model env es [Double]
-dirichlet' xs = Model $ do
-  call (Dist (DirichletDist xs) Nothing Nothing)
+dirichlet' :: (Has (ObsReader env :+: Draw) sig m)
+  => [Double] -- ^ concentration parameters
+  -> m [Double]
+dirichlet' xs = draw (DirichletDist xs) Nothing Nothing
 
-discrete :: forall env es x. Observable env x Int =>
-     [Double]
+discrete :: forall env sig m x. (Observable env x Int, Has (ObsReader env :+: Draw) sig m)
+  => [Double]
   -> ObsVar x
-  -> Model env es Int
-discrete ps field = Model $ do
-  let tag = Just $ varToStr field
-  maybe_y <- ask @env field
-  call (Dist (DiscreteDist ps) maybe_y tag)
+  -> m Int
+discrete ps = envDraw @env (DiscreteDist ps)
 
-discrete' ::
-     [Double]         -- ^ list of @n@ probabilities
-  -> Model env es Int -- ^ integer index from @0@ to @n - 1@
-discrete' ps = Model $ do
-  call (Dist (DiscreteDist ps) Nothing Nothing)
+discrete' :: (Has (ObsReader env :+: Draw) sig m)
+  => [Double]         -- ^ list of @n@ probabilities
+  -> m Int -- ^ integer index from @0@ to @n - 1@
+discrete' ps = draw (DiscreteDist ps) Nothing Nothing
 
-categorical :: forall env es a x. (Eq a, Show a, OpenSum.Member a PrimVal, Observable env x a) =>
-     [(a, Double)]
+categorical :: forall env sig m x a. (Eq a, Show a, OpenSum.Member a PrimVal, Observable env x a, Has (ObsReader env :+: Draw) sig m)
```

Comment:
It seems like in fused-effects multiple `Has` statements are preffered over using effect sums like this, is there a particular reason you've combined them into one Has statement?